src/librbd/migration/RawSnapshot.cc

```diff
@@ -205,10 +206,16 @@ void RawSnapshot<I>::list_snap(io::Extents&& image_extents,
   auto cct = m_image_ctx->cct;
   ldout(cct, 20) << "image_extents=" << image_extents << dendl;
 
-  // raw does support sparse extents so list the full IO extent as a delta
-  for (auto& [image_offset, image_length] : image_extents) {
-    sparse_extents->insert(image_offset, image_length,
-                           {io::SPARSE_EXTENT_STATE_DATA, image_length});
+  std::shared_ptr<NBDStream<I>> nbd_stream = std::dynamic_pointer_cast<NBDStream<I>>(m_stream);
+  if (nbd_stream != NULL) {
+    nbd_stream->list_snap(std::move(image_extents), sparse_extents, on_finish);
```

Comment:
This looks hackish indeed.

I suppose you could add `list_snap` method to `SnapshotInterface`, so you would just call `m_stream>list_snap` here. That would mean that you would either add "raw" implementation to all other stream implementations, or had it as a default in `SnapshotInterface` (i.e. `list_snap` would not be a pure virtual function, which probably "breaks" interface meaning, though I would personally not mind, not sure about others).